### DISCUSSION TITLE
【求助】iOS PacketTunnel 进程崩溃

### DISCUSSION THREAD

**Discussion by arror** [eycorsican/leaf#265] [Q&A] [answered]
iOS PacketTunnel 进程崩溃。

Branch：master

使用 make ios编译 libleaf.a

Cargo.toml：

[features]
default = [
    "leaf/config-json",
    "leaf/inbound-tun",
    "leaf/outbound-socks"
]

leaf-ffi/src/libs.rs：新增接口

```rust
#[no_mangle]
pub extern "C" fn leaf_run_with_config(rt_id: u16, config: *const c_char) -> i32 {
    if let Ok(_config) = unsafe { CStr::from_ptr(config).to_str() } {
        let opts = leaf::StartOptions {
            config: leaf::Config::Str(_config.to_string()),
            #[cfg(feature = "auto-reload")]
            auto_reload: false,
            runtime_opt: leaf::RuntimeOption::SingleThread,
        };
        if let Err(e) = leaf::start(rt_id, opts) {
            return to_errno(e);
        }
        ERR_OK
    } else {
        ERR_CONFIG
    }
}
```

接口调用：
```swift
leaf_run_with_config(987, config.cString(using: .utf8))
```

配置文件：
```json
{
    "log": {
        "level": "trace"
    },
    "inbounds": [
        {
            "protocol": "tun",
            "settings": {
                "fd": \(fd)
            },
            "tag": "tun"
        }
    ],
    "outbounds": [
        {
            "protocol": "socks",
            "settings": {
                "address": "127.0.0.1",
                "port": 8080
            },
            "tag": "clash"
        }
    ]
}
```

输出日志：

\^[[37m[2022-04-17 12:24:24][\^[[37mDEBUG\^[[0m\^[[37m] default handler [clash]\^[[0m
\^[[90m[2022-04-17 12:24:24][\^[[90mTRACE\^[[0m\^[[90m] added handler [clash]\^[[0m
\^[[90m[2022-04-17 12:24:24][\^[[90mTRACE\^[[0m\^[[90m] added runtime 987\^[[0m

崩溃堆栈：

  - 0 : "0   ???                                 0x0000000104fa89a0 0x0 + 4378495392"
  - 1 : "1   ???                                 0x0000000104fa8a2c 0x0 + 4378495532"
  - 2 : "2   PacketTunnel-iOS                    0x0000000102b55010 PacketTunnel-iOS + 5771280"
  - 3 : "3   libsystem_c.dylib                   0x00000001c6ca451c abort + 168"
  - 4 : "4   PacketTunnel-iOS                    0x0000000102976448 netif_add + 996"
  - 5 : "5   PacketTunnel-iOS                    0x0000000102974898 netif_init + 84"
  - 6 : "6   PacketTunnel-iOS                    0x00000001029747d8 lwip_init + 28"
  - 7 : "7   PacketTunnel-iOS                    0x0000000102b52658 _ZN3std4sync4once4Once10call_inner17hd5a75708fb5e7ae7E + 744"
  - 8 : "8   PacketTunnel-iOS                    0x00000001029f8ab4 _ZN97_$LT$core..future..from_generator..GenFuture$LT$T$GT$$u20$as$u20$core..future..future..Future$GT$4poll17h5c6d3d814f45d00fE + 2112"
  - 9 : "9   PacketTunnel-iOS                    0x0000000102a06fbc _ZN97_$LT$core..future..from_generator..GenFuture$LT$T$GT$$u20$as$u20$core..future..future..Future$GT$4poll17hac81694ea5332737E + 3016"
  - 10 : "10  PacketTunnel-iOS                    0x0000000102a7df78 _ZN103_$LT$futures_util..future..select_all..SelectAll$LT$Fut$GT$$u20$as$u20$core..future..future..Future$GT$4poll17h452499cb31f403eeE + 76"
  - 11 : "11  PacketTunnel-iOS                    0x0000000102a7a2c8 _ZN5tokio7runtime7Runtime8block_on17hcae03aeaceab0421E + 3260"
  - 12 : "12  PacketTunnel-iOS                    0x0000000102a78e3c _ZN4leaf5start17ha606316defa5e987E + 12236"
  - 13 : "13  PacketTunnel-iOS                    0x0000000102998c54 leaf_run_with_config + 172"
  - 14 : "14  PacketTunnel-iOS                    0x00000001025e0afc $s16PacketTunnel_iOS0aB8ProviderC05startB07optionsySDySSSo8NSObjectCGSg_tYaKFyyYbcfU2_ + 1540"
  - 15 : "15  PacketTunnel-iOS                    0x00000001025e0bec $sIegh_IeyBh_TR + 52"
  - 16 : "16  libdispatch.dylib                   0x00000001bb826e68 69DAC9C7-DDAF-318E-B068-BE596D142A2F + 7784"
  - 17 : "17  libdispatch.dylib                   0x00000001bb828a2c 69DAC9C7-DDAF-318E-B068-BE596D142A2F + 14892"
  - 18 : "18  libdispatch.dylib                   0x00000001bb82bb44 69DAC9C7-DDAF-318E-B068-BE596D142A2F + 27460"
  - 19 : "19  libdispatch.dylib                   0x00000001bb83a164 69DAC9C7-DDAF-318E-B068-BE596D142A2F + 86372"
  - 20 : "20  libdispatch.dylib                   0x00000001bb83a96c 69DAC9C7-DDAF-318E-B068-BE596D142A2F + 88428"
  - 21 : "21  libsystem_pthread.dylib             0x000000022cb46080 _pthread_wqthread + 228"
  - 22 : "22  libsystem_pthread.dylib             0x000000022cb45e5c start_wqthread + 8"

**Comment by LukkaVog** [comment]
同崩溃：
```

libsystem_c.dylib`__abort:
    0x1a2dd4830 <+0>:   adrp   x8, 284433
    0x1a2dd4834 <+4>:   ldr    x9, [x8, #0x8e8]
    0x1a2dd4838 <+8>:   cbnz   x9, 0x1a2dd4848           ; <+24>
    0x1a2dd483c <+12>:  adrp   x9, 8
    0x1a2dd4840 <+16>:  add    x9, x9, #0x2d8            ; "__abort() called"
    0x1a2dd4844 <+20>:  str    x9, [x8, #0x8e8]
    0x1a2dd4848 <+24>:  sub    sp, sp, #0x40
    0x1a2dd484c <+28>:  stp    x20, x19, [sp, #0x20]
    0x1a2dd4850 <+32>:  stp    x29, x30, [sp, #0x30]
    0x1a2dd4854 <+36>:  add    x29, sp, #0x30
    0x1a2dd4858 <+40>:  str    xzr, [sp, #0x10]
    0x1a2dd485c <+44>:  add    x8, sp, #0x10
    0x1a2dd4860 <+48>:  add    x19, x8, #0x8
    0x1a2dd4864 <+52>:  movi   d0, #0x000000ffffffff
    0x1a2dd4868 <+56>:  str    d0, [sp, #0x18]
    0x1a2dd486c <+60>:  add    x1, sp, #0x10
    0x1a2dd4870 <+64>:  orr    w0, wzr, #0x6
    0x1a2dd4874 <+68>:  mov    x2, #0x0
    0x1a2dd4878 <+72>:  bl     0x1a2d725b8               ; sigaction
    0x1a2dd487c <+76>:  ldr    w8, [sp, #0x18]
    0x1a2dd4880 <+80>:  and    w8, w8, #0xffffffdf
    0x1a2dd4884 <+84>:  bl     0x1a2db4a64               ; OUTLINED_FUNCTION_0
    0x1a2dd4888 <+88>:  orr    w0, wzr, #0x1
    0x1a2dd488c <+92>:  bl     0x1a2e8d068               ; symbol stub for: __pthread_workqueue_setkill
    0x1a2dd4890 <+96>:  bl     0x1a2db4a80               ; OUTLINED_FUNCTION_1
    0x1a2dd4894 <+100>: bl     0x1a2e8c418               ; pthread_self
    0x1a2dd4898 <+104>: orr    w1, wzr, #0x6
    0x1a2dd489c <+108>: bl     0x1a2e8d350               ; symbol stub for: pthread_kill
    0x1a2dd48a0 <+112>: mov    w0, #0x2710
    0x1a2dd48a4 <+116>: bl     0x1a2d8dfec               ; usleep$NOCANCEL
    0x1a2dd48a8 <+120>: orr    w8, wzr, #0xffffffe7
    0x1a2dd48ac <+124>: str    w8, [sp, #0x18]
    0x1a2dd48b0 <+128>: orr    w0, wzr, #0x3
    0x1a2dd48b4 <+132>: mov    x1, x19
    0x1a2dd48b8 <+136>: mov    x2, #0x0
    0x1a2dd48bc <+140>: bl     0x1a2dd5aa8               ; symbol stub for: sigprocmask
->  0x1a2dd48c0 <+144>: brk    #0x1
```

大佬可以加下非官网tg群讨论学习：
https://t.me/joinchat/Hi8zv7Xzipi04Iq8

ref: https://github.com/eycorsican/leaf/discussions/156

**Reply by LukkaVog** [reply]
另外我这里用默认leaf_run后，有这些日志，后面就没有了：
```
[37m[2022-04-17 14:54:19][[37mDEBUG[0m[37m] default handler [Trojan][0m
[90m[2022-04-17 14:54:19][[90mTRACE[0m[90m] added handler [Trojan_tls_xxx][0m
[90m[2022-04-17 14:54:19][[90mTRACE[0m[90m] added handler [Trojan_trojan_xxx][0m
[90m[2022-04-17 14:54:19][[90mTRACE[0m[90m] added handler [DirectLan][0m
[90m[2022-04-17 14:54:19][[90mTRACE[0m[90m] added handler [Trojan] with actors: Trojan_tls_xxx,Trojan_trojan_xxx[0m
[90m[2022-04-17 14:54:19][[90mTRACE[0m[90m] added runtime 79[0m
[37m[2022-04-17 14:54:19][[32mINFO[0m[37m] inbound listening tcp 127.0.0.1:10793[0m
```

conf:
```
confString: # 全局模式

[General]
loglevel = trace
logoutput = /private/var/mobile/Containers/Shared/AppGroup/XXXXXXXXXX-XXX-XXXX-XXXX-XXXXXXXXX/leaf-run.log
dns-server = 114.114.114.114, 223.5.5.5
routing-domain-resolve = true
always-real-ip = apple.com
tun-fd = 4

socks-interface = 127.0.0.1
socks-port = 10793

[Proxy]
Trojan = trojan, xxx.xx.xx.xx, xxx, password=xxxxxx, sni=i.apple.com
DirectLan = direct

[Rule]
# LAN
DOMAIN-SUFFIX, local, DirectLan
DOMAIN-SUFFIX, test, DirectLan
IP-CIDR, 127.0.0.0/8, DirectLan
IP-CIDR, 172.16.0.0/12, DirectLan
IP-CIDR, 192.168.0.0/16, DirectLan
IP-CIDR, 10.0.0.0/8, DirectLan
IP-CIDR, 17.0.0.0/8, DirectLan
IP-CIDR, 100.64.0.0/10, DirectLan
IP-CIDR, 224.0.0.0/4, DirectLan
IP-CIDR6, fe80::/10, DirectLan

FINAL, Trojan
```

**Reply by arror** [reply]
在leaf_run的位置，打印一下崩溃的线程堆栈信息，看看和我的崩溃信息是不是一样的。

看起来lwip初始化的时候有问题。

我用go-tun2socks没有问题。

**Comment by arror** [comment, marked as answer]
Close!!!

初始化顺序导致。

**Comment by 594810899** [comment]
> iOS PacketTunnel 进程崩溃。
> 
> Branch：master
> 
> 使用 make ios编译 libleaf.a
> 


请问一下使用make ios 命令怎么编译的libleaf.a  有没有具体一些的操作步骤？

**Reply by arror** [reply]
配置好rust环境，rust-lipo，以及target，在项目根目录下，执行 make ios，实际上执行的是Makefile中对应的命令

**Reply by arror** [reply]
去掉 --no-default-features --features "default-openssl"试试

**Reply by arror** [reply]
应该是，先确保aarch64-apple-ios添加成功在编译吧

**Reply by 594810899** [reply]
>我电脑执行这个添加命令的时候 卡主不动了  执行N次了都不成功，  我用这上面编译好的那个[Release v0.3.1](https://github.com/eycorsican/leaf/releases/tag/v0.3.1) libleaf.a 库， iOS端可以使用吗？ 我看这上面最近几个编译的Release都没有iOS的

**Reply by arror** [reply]
应该可以。不过不建议使用旧版本。如果你只想用leaf提供tun2socks能力，可以试试这个：https://github.com/AppleClash/T2SKit ，我基于leaf编译的Swift Package。

**Reply by 594810899** [reply]
>非常感谢！ 我正是需要leaf的tun2socks功能，我看了这个里面"outbounds": [
                {
                    "protocol": "socks",
                    "settings": {
                        "address": "\(host)",
                        "port": \(port)
                    },
                    "tag": "clash"
                }
            ] 是sock.  是否支持trojan？

**Reply by arror** [reply]
tag只是标签。只要是socks就可以。不支持auth。

**Reply by arror** [reply]
数据流向大概是这样的 app --> tun --> leaf --> trojan client --> trojan server。
所以 inbound 是 tun，outbound 是 Trojan client 监听的 socks。

**Reply by 594810899** [reply]
> 数据流向大概是这样的 app --> tun --> leaf --> trojan client --> trojan server。
> 所以 inbound 是 tun，outbound 是 Trojan client 监听的 socks。


大佬，终于用最新的leaf 编译出来 libleaf.a库  可是没有看到leaf.h文件， 这个文件是需要另外操作还是隐藏在哪里了？

**Reply by arror** [reply]
cbindgen --config leaf-ffi/cbindgen.toml leaf-ffi/src/lib.rs > target/universal/release/leaf.h

**Reply by arror** [reply]
T2SKit不支持Trojan outbound，如果需要，https://github.com/AppleClash/tun2socks/blob/930fb32b7067b6056ec4eb84fa18701da66d29f1/tun2socks/Cargo.toml#L16  default里添加 "leaf/outbound-trojan"，重新编译。
```json
{
  "log": {
    "level": "warn"
  },
  "inbounds": [
    {
      "protocol": "tun",
      "settings": {
        "fd": (fd)
      },
      "tag": "tun"
    }
  ],
  "outbounds": [
    {
      "protocol": "trojan",
      "settings": {
        "address": "212.50.244.81",
        "password":"myncic_123456",
        "port": 6666
      },
      "tag": "trojan_out"
    }
  ]
}
```

**Reply by arror** [reply]
Telegram
